### PR TITLE
Add filter method alternative to linesearch

### LINKED ISSUE #31: Filter line-search
Armijo is not enough

We can start looking here: https://epubs.siam.org/doi/pdf/10.1137/S1052623403426556

See https://github.com/Simple-Robotics/proxnlp/issues/7

### PR DESCRIPTION
**Pull request by edantec**
This PR:

- Adds a pair filter option to manage step acceptance in place of linesearch
- Provides a python example of a walking OCP on Talos. 

Resolves #31

### CODE REVIEW

**Comment by ManifoldFR** [comment]
Thanks ! I'll do some editing to this PR if you don't mind

**Comment by edantec** [comment]
In commit [a7dee9f] I point out a circular include of enums.hpp in solver-proxddp.hpp

**Comment by jcarpent** [comment]
> In commit [[a7dee9f](https://github.com/Simple-Robotics/aligator/commit/a7dee9f5b2d567e51176939d9a077ed7c45ec0c6)] I point out a circular include of enums.hpp in solver-proxddp.hpp

It should not be a problem is there is the classic safe guard or the new pragma once.
Is it the case?

**Review by ManifoldFR — COMMENTED** [review]
I have a few comments

File: bindings/python/src/module.cpp [hunk 1]
```diff
@@ -35,6 +35,11 @@ void exposeEnums() {
       .value("HESSIAN_EXACT", HessianApprox::EXACT)
       .value("HESSIAN_GAUSS_NEWTON", HessianApprox::GAUSS_NEWTON)
       .export_values();
+
+  bp::enum_<StepAcceptanceMode>("StepAcceptanceMode", "Step acceptance mode.")
+      .value("LINESEARCH", StepAcceptanceMode::LINESEARCH)
+      .value("FILTER", StepAcceptanceMode::FILTER)
+      .export_values();
```
[the remark this answers is not in the document]
Comment (jcarpent) [reply bindings/python/src/module.cpp:42]: `StepAcceptanceMode` should be renamed too and becomes something like `StepAcceptanceStrategy`

**Comment by ManifoldFR** [comment]
The CI - OSX/Linux via Conda build is failing because we're using the features for `std::pair` from the [new eigenpy 3.2.0](https://github.com/stack-of-tasks/eigenpy/releases/tag/v3.2.0) and the conda recipe is still on the previous version of eigenpy. Switching to `std::array<_, 2>` also won't do it for the same reason

We can do two things:
* upgrade the eigenpy requirement to 3.2.0 for the entire library (fine by me) and the conda recipe
* replace `std::pair<Scalar, Scalar>` by a simple `struct`
* manually expose the pair type (with an `eigenpy::register_symbolic_link` safeguard)

Both options seem good to me.

**Comment by ManifoldFR** [comment]
By the way @edantec please add yourself as a contributor to the README :)

File: examples/utils/__init__.py [hunk 2]
```diff
@@ -201,6 +201,14 @@ def load_talos_upper_body():
     return red_bot
 
 
+def load_talos():
```
[the remark this answers is not in the document]
Comment (edantec) [reply examples/utils/__init__.py]: Name changed to load_talos_no_wristhead

File: include/aligator/solvers/proxddp/solver-proxddp.hpp [hunk 3]
```diff
@@ -327,6 +332,8 @@ template <typename _Scalar> struct SolverProxDDP {
   Scalar rho_penal_ = rho_init;
   /// Linesearch function
   LinesearchType linesearch_;
+  Filter filter_;
+  Scalar beta_ = 1e-5;
```
[the remark this answers is not in the document]
Comment (edantec) [reply include/aligator/solvers/proxddp/solver-proxddp.hpp]: beta is removed from the header, and initialized to 0 in the .hxx

File: bindings/python/src/module.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (edantec) [reply bindings/python/src/module.cpp:42]: Done, we use StepAcceptanceStrategy now.

**Comment by jcarpent** [comment]
> The CI - OSX/Linux via Conda build is failing because we're using the features for `std::pair` from the [new eigenpy 3.2.0](https://github.com/stack-of-tasks/eigenpy/releases/tag/v3.2.0) and the conda recipe is still on the previous version of eigenpy

A new version of Pinocchio with the latest eigenpy release is on its way (see https://github.com/conda-forge/pinocchio-feedstock/pull/107).

**Comment by ManifoldFR** [comment]
I ended up bumping the required eigenpy version to 3.2.0.

**Comment by jcarpent** [comment]
@edantec What is the status of this PR?

**Comment by ManifoldFR** [comment]
New packages for proxsuite-nlp 0.3.2 (using eigenpy 3.2.0) got shipped, I'll trigger the CI jobs again.

**Comment by ManifoldFR** [comment]
@edantec Please update the changelog

**Review by jcarpent — DISMISSED** [review]

**Review by ManifoldFR — COMMENTED** [review]
Oh wow, I never noticed this.

**Comment by edantec** [comment]
The memory error on test_solver.py has been fixed. As for the memory error of test_integrators.py, I have no idea why but it seems to have disappeared according to Joris.

**Comment by ManifoldFR** [comment]
> The memory error on test_solver.py has been fixed. As for the memory error of test_integrators.py, I have no idea why but it seems to have disappeared according to Joris.

It's the magic of ~Christmas~C++

**Review by ManifoldFR — APPROVED** [review]

**Pull request merged**